src/network/networkd-sysctl.c

```diff
@@ -5,17 +5,189 @@
+
+        rootcg = cg_path_open(SYSTEMD_CGROUP_CONTROLLER, "/");
+        if (rootcg < 0)
+                return log_warning_errno(rootcg, "Failed to open cgroup: %m");
+
+        obj = sysctl_monitor_bpf__open_and_load();
+        if (!obj)
+                return log_warning_errno(errno, "Failed to load sysctl monitor BPF program: %m");
+
+        cgroup_fd = cg_path_open(SYSTEMD_CGROUP_CONTROLLER, "/system.slice/systemd-networkd.service");
+        if (cgroup_fd < 0)
+                return log_warning_errno(cgroup_fd, "Failed to open cgroup: %m");
+
+        if (sym_bpf_map_update_elem(sym_bpf_map__fd(obj->maps.cgroup_map), &idx, &cgroup_fd, BPF_ANY))
+                return log_error_errno(errno, "Failed to update cgroup map: %m");
+
+        sysctl_link = sym_bpf_program__attach_cgroup(obj->progs.sysctl_monitor, rootcg);
+        r = bpf_get_error_translated(sysctl_link);
+        if (r < 0)
+                return log_warning_errno(r, "Failed to attach sysctl monitor BPF program to cgroup: %m");
+
+        sysctl_buffer = sym_ring_buffer__new(
+                        sym_bpf_map__fd(obj->maps.written_sysctls),
+                        sysct_write_event, &manager->sysctl_shadow, NULL);
+        if (!sysctl_buffer)
+                return log_warning_errno(errno, "Failed to create ring buffer: %m");
+
+        r = sd_event_add_io(manager->event, &manager->sysctl_event_source,
+                        sym_ring_buffer__epoll_fd(sysctl_buffer), EPOLLIN, on_ringbuf_io, sysctl_buffer);
+        if (r < 0)
+                return log_warning_errno(r, "Failed to watch sysctl event ringbuffer: %m");
+
+        manager->sysctl_link = TAKE_PTR(sysctl_link);
+        manager->sysctl_skel = TAKE_PTR(obj);
+        manager->sysctl_buffer = TAKE_PTR(sysctl_buffer);
+        manager->cgroup_fd = TAKE_FD(cgroup_fd);
+
+        return 0;
+}
+
+int sysctl_remove_monitor(Manager *manager) {
+        _cleanup_close_ int rootcg = -EBADF;
+        int r;
+
+        assert(manager);
+
+        r = dlopen_bpf();
+        if (r < 0)
+                return r;
+
+        rootcg = cg_path_open(SYSTEMD_CGROUP_CONTROLLER, "/");
+        if (rootcg < 0)
+                return r;
+
+        if (manager->sysctl_event_source)
+                manager->sysctl_event_source = sd_event_source_disable_unref(manager->sysctl_event_source);
+
+        if (manager->sysctl_buffer) {
+                sym_ring_buffer__free(manager->sysctl_buffer);
+                manager->sysctl_buffer = NULL;
+        }
+
+        if (manager->sysctl_link) {
+                sym_bpf_link__destroy(manager->sysctl_link);
+                manager->sysctl_link = NULL;
+        }
+
+        if (manager->sysctl_skel) {
+                sysctl_monitor_bpf__destroy(manager->sysctl_skel);
+                manager->sysctl_skel = NULL;
+        }
+
+        if (manager->cgroup_fd >= 0)
+                manager->cgroup_fd = safe_close(manager->cgroup_fd);
```

Comment:
and here too. the conditionalization is redundant. safe_close() happily accepts fds < 0 and is a NOP then.